# Thông tin đồ án

## Thành viên

|   MSSV   |      Họ tên     |   GitHub  |
|:--------:|:---------------:|:---------:|
| 18120120 | Trần Chí Hào    | haotrch   |
| 18120147 | Trần Quốc Thắng | tqthangbl |

## Link thùng chứa GitHub

[https://github.com/tqthangbl/P4DS-FinalProject](https://github.com/tqthangbl/P4DS-FinalProject)

# Môi trường code



- Code trên môi trường Linux
- Dùng phiên bản các package như trong file "min_ds-env.yml"


In [1]:
import sys
sys.executable

'/home/family-u/.conda/envs/min_ds-env/bin/python'

# Import các thư viện cần thiết

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

# Thu thập dữ liệu

**Dữ liệu của bạn là về chủ đề gì và bạn lấy từ nguồn nào?**

Bộ dữ liệu mà nhóm sử dụng có tên là "Google Play Store Apps", được lấy từ [Kaggle](https://www.kaggle.com/lava18/google-play-store-apps), chứa thông tin về hơn 10000 ứng dụng trên Google Play Store. Thời điểm bộ dữ liệu được đăng lên Kaggle là 2018-09-05, lần cập nhật cuối là vào 2019-02-03.

Có 2 file dữ liệu:
1. `googleplaystore.csv`: Chứa các thông tin chi tiết về các ứng dụng thu thập được từ Google Play
2. `googleplaystore_user_reviews`: Chứa thông tin về 100 đánh giá liên quan nhất cho mỗi ứng dụng thu thập được từ Google Play

**Người ta có cho phép bạn dùng dữ liệu như này không? Ví dụ, bạn có thể xem thử license của dữ liệu là gì**


License của dữ liệu là Creative Commons Attribution 3.0 Unported, theo như được nói ở [đây](https://creativecommons.org/licenses/by/3.0/), người ta cho phép:

1. Copy và chia sẻ dữ liệu này dưới bất kì hình thức nào
2. Sử dụng bộ dữ liệu này để xây dựng bất cứ thứ gì, cho bất cứ mục đích nào, kể cả thương mại

**Người ta thu thập dữ liệu như thế nào?**

Dữ liệu được tác giả crawl từ Google Play Store.


# Khám phá dữ liệu (thường đan xen với tiền xử lý)

* Dữ liệu có bao nhiêu dòng và bao nhiêu cột?
* Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa
khác nhau không?
* Dữ liệu có các dòng bị lặp không?
* Mỗi cột có ý nghĩa gì?
* Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu
dữ liệu chưa phù hợp để có thể xử lý tiếp không?
* Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá
trị được phân bố như thế nào?
    * Số-lượng/tỉ-lệ các giá trị thiếu?
    * Min? max? → Có gì bất thường không?
* Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical),
các giá trị được phân bố như thế nào?
    * Số-lượng/tỉ-lệ các giá trị thiếu?
    * Số lượng các giá trị khác nhau? Show một vài giá trị
→ Có gì bất thường không?

## Đọc dữ liệu từ file csv vào dataframe

Ở đây nhóm khám phá dữ liệu trong file `googleplaystore.csv` trước. Dữ liệu đọc từ file sẽ lưu vào dataframe `app_df`.

In [3]:
app_df = pd.read_csv('Data/googleplaystore.csv')
app_df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

## Dữ liệu có bao nhiêu dòng và bao nhiêu cột?

In [4]:
num_rows = len(app_df)
num_cols = len(app_df.columns)

print(f'[+] num_rows = {num_rows}')
print(f'[+] num_cols = {num_cols}')

[+] num_rows = 10841
[+] num_cols = 13


Dữ liệu có 10841 dòng và 13 cột.

## Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

In [5]:
app_df.sample(5)

App       Category  \
3563                          aCalendar - Android Phone   PRODUCTIVITY   
2118  Disney Magic Kingdoms: Build Your Own Magical ...         FAMILY   
279                                     AdWords Express       BUSINESS   
4636                                             /u/app  COMMUNICATION   
506                                               Howlr         DATING   

      Rating Reviews                Size     Installs  Type Price  \
3563     4.4   80119  Varies with device  10,000,000+  Free     0   
2118     4.3  472584                 91M  10,000,000+  Free     0   
279      4.1    7149                 11M   1,000,000+  Free     0   
4636     4.7     573                 53M      10,000+  Free     0   
506      3.9     105                 27M       5,000+  Free     0   

     Content Rating                         Genres    Last Updated  \
3563       Everyone                   Productivity   June 25, 2018   
2118       Everyone  Simulation;Action & Adventure  August 2, 2018   
279        Everyone                       Business   July 31, 2018   
4636     Mature 17+                  Communication    July 3, 2018   
506      Mature 17+                         Dating   June 19, 2018   

             Current Ver         Android Ver  
3563  Varies with device  Varies with device  
2118              3.2.0h        4.0.3 and up  
279              2.6.158          4.0 and up  
4636               4.2.4          4.1 and up  
506                1.0.4          4.4 and up

Theo như mô tả của tác giả, mỗi dòng trong dữ liệu chứa thông tin của 1 ứng dụng trên Play Store. Bằng việc quan sát ngẫu nhiên các dòng trong dữ liệu nhiều lần, có vẻ không có dòng nào mang ý nghĩa khác. 

## Mỗi cột có ý nghĩa gì?

Theo như mô tả trên Kaggle, bộ dữ liệu gồm 13 cột với ý nghĩa từng cột như sau:

- **App**: Application name
- **Category**: Category the app belongs to 
- **Rating**: Overall user rating of the app (as when scraped)
- **Reviews**: Number of user reviews for the app (as when scraped)
- **Size**: Size of the app (as when scraped)
- **Installs**: Number of user downloads/installs for the app (as when scraped)
- **Type**: Paid or Free
- **Price**: Price of the app (as when scraped)
- **Content Rating**: Age group the app is targeted at - Children / Mature 21+ / Adult
- **Genres**: An app can belong to multiple genres (apart from its main category). For eg, a musical family game will belong to Music, Game, Family genres.

3 cột còn lại không có mô tả trên Kaggle, nhưng dựa vào tên cột và quan sát các dòng ngẫu nhiên, có thể hiểu ý nghĩa của chúng như sau:

- **Last Updated**: Lần cuối ứng dụng được cập nhật trên Play Store
- **Current Ver**: Phiên bản hiện tại của ứng dụng trên Play Store
- **Android Ver**: Các phiên bản Android phù hợp để cài đặt ứng dụng

## Dữ liệu có các dòng bị lặp không?

In [6]:
have_duplicated_rows = True
dup_df = app_df[app_df.duplicated()]
if len(dup_df) == 0:
    have_duplicated_rows = False
    
print(f'[+] have_duplicated_rows = {have_duplicated_rows}')

[+] have_duplicated_rows = True


### Vấn đề cần tiền xử lý
- Một số dòng trong dữ liệu bị lặp, điều này có thể dẫn đến khi tính toán (đếm, tính tỉ lệ). Tuy là Play Store cho phép 2 ứng dụng có thể trùng tên, nhưng rất khó để 2 ứng dụng có thể trùng toàn bộ thông tin trong 12 cột còn lại. Do đó những dòng bị lặp này là của cùng 1 ứng dụng, vì vậy cần phải loại bỏ chúng.

---

# Tiền xử lý

## Loại bỏ các dòng bị lặp

Các dòng bị lặp được lưu trong dataframe `dup_df`. Đầu tiên xem thử 1 vài giá trị của dataframe này.

In [7]:
dup_df.sample(5)

App      Category  Rating  \
828                                  Learn 50 languages     EDUCATION     4.4   
785    Quizlet: Learn Languages & Vocab with Flashcards     EDUCATION     4.6   
8630  Any.do: To-do list, Calendar, Reminders & Planner  PRODUCTIVITY     4.5   
2227                         Thomas & Friends: Race On!        FAMILY     4.1   
2367  mySugr: the blood sugar tracker made just for you       MEDICAL     4.6   

     Reviews                Size     Installs  Type Price Content Rating  \
828    55256                 14M   5,000,000+  Free     0       Everyone   
785   211856  Varies with device  10,000,000+  Free     0       Everyone   
8630  298854  Varies with device  10,000,000+  Free     0       Everyone   
2227   29319                 26M   5,000,000+  Free     0       Everyone   
2367   21189                 36M   1,000,000+  Free     0       Everyone   

                         Genres    Last Updated         Current Ver  \
828                   Education   June 19, 2018              10.9.1   
785                   Education  August 1, 2018  Varies with device   
8630               Productivity  August 5, 2018  Varies with device   
2227  Casual;Action & Adventure    May 18, 2018                 2.4   
2367                    Medical  August 6, 2018              3.52.1   

             Android Ver  
828           4.0 and up  
785   Varies with device  
8630  Varies with device  
2227          4.4 and up  
2367          5.0 and up

Tiến hành loại bỏ các dòng bị lặp trong `app_df` và cập nhật lại biến `num_rows` thể hiện số dòng của dữ liệu.

In [8]:
app_df = app_df.drop_duplicates()
num_rows = len(app_df)

print(f'[+] num_rows = {num_rows}')

[+] num_rows = 10358


Số dòng của dữ liệu sau khi loại bỏ các dòng bị lặp, chỉ còn lại 10358 dòng.

---

# Khám phá dữ liệu (tiếp tục)

## Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

Trước hết lấy ra `dtypes` của các cột, lưu vào series `col_dtypes`, và in ra để quan sát.

In [9]:
col_dtypes = app_df.dtypes

col_dtypes

App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price              object
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

Cột `Rating` thể hiện đánh giá của người dùng cho ứng dụng, là 1 con số nên kiểu `float64` là hợp lý. Các cột còn lại đang có kiểu `object`. Kiểu `object` trong Pandas thường là chuỗi, nhưng cũng có thể là một đối tượng bất kỳ. Do đó, trong các cột thuộc kiểu `object`, có thể chứa những giá trị thuộc các kiểu dữ liệu khác nhau. Ta phải kiểm tra điều này bằng cách: với mỗi cột, áp dụng lệnh `type` lên từng phần tử trong cột đó để xem kiểu dữ liệu thật sự của chúng. Hàm `open_object_dtype` sẽ nhận input là 1 series `s`, trả về 1 set là tập các kiểu dữ liệu có trong series đó. Khi xem kết quả hàm `open_object_dtype` của từng cột, ta cũng xem luôn cột đó mang các giá trị phân biệt nào để xem có điều gì bất thường hay không.

In [10]:
def open_object_dtype(s):
    return set(s.apply(type).unique())

for col_name in col_dtypes.index:
    if col_dtypes[col_name] == 'object':
        print(f'[+] Column {col_name} dtypes: ')
        print(open_object_dtype(app_df[col_name]))
        print(app_df[col_name].unique())
        print()

[+] Column App dtypes: 
{<class 'str'>}
['Photo Editor & Candy Camera & Grid & ScrapBook' 'Coloring book moana'
 'U Launcher Lite – FREE Live Cool Themes, Hide Apps' ...
 'Parkinson Exercices FR' 'The SCP Foundation DB fr nn5n'
 'iHoroscope - 2018 Daily Horoscope & Astrology']

[+] Column Category dtypes: 
{<class 'str'>}
['ART_AND_DESIGN' 'AUTO_AND_VEHICLES' 'BEAUTY' 'BOOKS_AND_REFERENCE'
 'BUSINESS' 'COMICS' 'COMMUNICATION' 'DATING' 'EDUCATION' 'ENTERTAINMENT'
 'EVENTS' 'FINANCE' 'FOOD_AND_DRINK' 'HEALTH_AND_FITNESS' 'HOUSE_AND_HOME'
 'LIBRARIES_AND_DEMO' 'LIFESTYLE' 'GAME' 'FAMILY' 'MEDICAL' 'SOCIAL'
 'SHOPPING' 'PHOTOGRAPHY' 'SPORTS' 'TRAVEL_AND_LOCAL' 'TOOLS'
 'PERSONALIZATION' 'PRODUCTIVITY' 'PARENTING' 'WEATHER' 'VIDEO_PLAYERS'
 'NEWS_AND_MAGAZINES' 'MAPS_AND_NAVIGATION' '1.9']

[+] Column Reviews dtypes: 
{<class 'str'>}
['159' '967' '87510' ... '603' '1195' '398307']

[+] Column Size dtypes: 
{<class 'str'>}
['19M' '14M' '8.7M' '25M' '2.8M' '5.6M' '29M' '33M' '3.1M' '28M' '12M

### Vấn đề cần tiền xử lý

Quan sát kết quả, khoan nói đến kiểu dữ liệu ở từng cột, ta thấy có một số điểm bất thường khi xem các giá trị phân biệt ở từng cột:
- Ở cột `Category`, có giá trị là `1.9`, có vẻ không phải là 1 giá trị phù hợp với ý nghĩa category
- Ở cột `Size`, có giá trị là `1,000+`, so với các giá trị còn lại trong cột này thường có format là 1 con số, theo sau đó là đơn vị (`k` hoặc `M`). Giá trị `1,000+` này có vẻ phù hợp với format của cột ngay sau đó là `Install` hơn.
- Ở cột `Installs`, có 1 giá trị là `Free`, không nói lên số lượt cài đặt của ứng dụng, giá trị này lại phù hợp với cột tiếp theo là `Type`, mang ý nghĩa ứng dụng đó thuộc loại miễn phí hay trả phí thì đúng hơn.

Từ đó ta đoán rằng có một dòng nào đó trong dữ liệu bị thiếu giá trị 1 số cột, do đó các giá trị của các cột còn lại bị dịch qua trái, nên giá trị ở một cột này lại nằm ở cột ngay trước đó. Đọc phần [https://www.kaggle.com/lava18/google-play-store-apps/discussion/164101](Discussion) trên Kaggle cũng có người gặp vấn đề này, họ nói rằng dòng 10472 bị thiếu giá trị. Nhưng ta vẫn cần tiền xử lý để tìm và sửa các dòng như vậy, cho chắc.

# Tiền xử lý

## Tìm các dòng có giá trị một số cột bị thiếu

Ta sử dụng thư viện `csv` để đọc file `googleplaystore.csv`, dòng đầu của file sẽ là tên các cột. Với mỗi dòng tiếp theo, nếu số cột trong dòng đó khác với số lượng cột thì đưa chỉ số của dòng vào list `bad_index`. Sau khi chạy code, `bad_index` sẽ chứa các chỉ số dòng có các cột bị thiếu.

In [11]:
import csv

csv_file = open('Data/googleplaystore.csv')
csv_reader = csv.reader(csv_file, delimiter=',')
row_id = -1
col_count = 0
bad_index = []

for row in csv_reader:
    if row_id == -1:
        col_count = len(row)
        print(f"CSV file has {col_count} columns")
    else:
        if len(row) != col_count:
            print(f"Row id = {row_id} has {len(row)} columns")
            bad_index.append(row_id)
    row_id += 1

CSV file has 13 columns
Row id = 10472 has 12 columns


## Thay thế các giá trị cột bị thiếu thành `NaN`

Có 1 dòng bị thiếu giá trị cột, là dòng có index 10472. Do dataframe `app_df` đã bị loại bỏ các dòng trùng, nên thứ tự các dòng trong file này không giống thứ tự trong file gốc. Vì vậy cần load lại `app_df`, và sửa đổi dòng bị thiếu. Sau đó mới loại bỏ lại các dòng bị trùng.

In [12]:
app_df = pd.read_csv('Data/googleplaystore.csv')

app_df.iloc[bad_index]

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

Ở dòng 10472, giá trị cột `Category` bị thiếu, nên ta dịch các giá trị bắt đầu từ cột `Category` sang bên phải rồi điền giá trị cho cột `Category` là `NaN`

In [13]:
cols_list = app_df.columns.tolist()
i = len(cols_list) - 1
while i > 1:
    app_df.loc[10472, cols_list[i]] = app_df.loc[10472, cols_list[i - 1]]
    i -= 1
app_df.loc[10472, 'Category'] = np.nan

app_df.iloc[10472]

App               Life Made WI-Fi Touchscreen Photo Frame
Category                                              NaN
Rating                                                1.9
Reviews                                              19.0
Size                                                 3.0M
Installs                                           1,000+
Type                                                 Free
Price                                                   0
Content Rating                                   Everyone
Genres                                                NaN
Last Updated                            February 11, 2018
Current Ver                                        1.0.19
Android Ver                                    4.0 and up
Name: 10472, dtype: object

Dữ liệu đã được sửa xong, bây giờ loại bỏ dòng bị lặp.

In [14]:
app_df = app_df.drop_duplicates()
num_rows = len(app_df)

print(f'[+] num_rows = {num_rows}')

[+] num_rows = 10358


# Khám phá dữ liệu (tiếp tục)

## Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không? (tiếp tục)

Ta tiếp tục liệt kê các kiểu dữ liệu khác nhau ở các cột có dtype là `object`.

In [15]:
def open_object_dtype(s):
    return set(s.apply(type).unique())

col_dtypes = app_df.dtypes

for col_name in col_dtypes.index:
    if col_dtypes[col_name] == 'object':
        print(f'[+] Column {col_name} dtypes: ')
        print(open_object_dtype(app_df[col_name]))
        print(app_df[col_name].unique())
        print()

[+] Column App dtypes: 
{<class 'str'>}
['Photo Editor & Candy Camera & Grid & ScrapBook' 'Coloring book moana'
 'U Launcher Lite – FREE Live Cool Themes, Hide Apps' ...
 'Parkinson Exercices FR' 'The SCP Foundation DB fr nn5n'
 'iHoroscope - 2018 Daily Horoscope & Astrology']

[+] Column Category dtypes: 
{<class 'str'>, <class 'float'>}
['ART_AND_DESIGN' 'AUTO_AND_VEHICLES' 'BEAUTY' 'BOOKS_AND_REFERENCE'
 'BUSINESS' 'COMICS' 'COMMUNICATION' 'DATING' 'EDUCATION' 'ENTERTAINMENT'
 'EVENTS' 'FINANCE' 'FOOD_AND_DRINK' 'HEALTH_AND_FITNESS' 'HOUSE_AND_HOME'
 'LIBRARIES_AND_DEMO' 'LIFESTYLE' 'GAME' 'FAMILY' 'MEDICAL' 'SOCIAL'
 'SHOPPING' 'PHOTOGRAPHY' 'SPORTS' 'TRAVEL_AND_LOCAL' 'TOOLS'
 'PERSONALIZATION' 'PRODUCTIVITY' 'PARENTING' 'WEATHER' 'VIDEO_PLAYERS'
 'NEWS_AND_MAGAZINES' 'MAPS_AND_NAVIGATION' nan]

[+] Column Rating dtypes: 
{<class 'str'>, <class 'float'>}
[4.1 3.9 4.7 4.5 4.3 4.4 3.8 4.2 4.6 3.2 4.0 nan 4.8 4.9 3.6 3.7 3.3 3.4
 3.5 3.1 5.0 2.6 3.0 1.9 2.5 2.8 2.7 1.0 2.9 2.3 2.2 1.

### Vấn đề cần tiền xử lý
- Cột `Category` thể hiện các danh mục mà ứng dụng thuộc về, để kiểu `str` là hợp lý, chỉ có giá trị `NaN` là `float`.
- Cột `Reviews` là số lượt đánh giá cho ứng dụng, đang ở kiểu `str` và `float`, giá trị này luôn là số nguyên nên cần chuyển về `int`.
- Cột `Type` thể hiện ứng dụng trả phí hoặc miễn phí, nên là kiểu `str`, chỉ có giá trị `NaN` là `float`.
- Cột `Price` thể hiện giá của ứng dụng, đang là `str`, có dạng dấu `$` sau đó là giá tiền, ngoại trừ giá `0`. Cần bỏ dấu `$` và chuyển cột này về `float`.
- Cột `Genres` thể hiện thể loại mà ứng dụng thuộc về, để kiểu `str` là hợp lý, chỉ có giá trị `NaN` là `float`.
- Cột `Last Updated` là thời điểm cập nhật gần nhất của ứng dụng trên Play Store, đang ở dạng `str`, cần chuyển về `datetime`.
- Cột `Current Ver` thể hiện phiên bản hiện tại ứng dụng trên Play Store, để kiểu `str` là hợp lý, chỉ có giá trị `NaN` là `float`.
- Cột `Android Ver` thể hiện phiên bản Android phù hợp để cài ứng dụng, để kiểu `str` là hợp lý, chỉ có giá trị `NaN` là `float`.

# Tiền xử lý

## Chuyển dtype của cột "Reviews" sang int

In [16]:
app_df = app_df.astype({'Reviews': int})

## Chuyển dtype của cột "Price" sang float

Cột này đang có dtype là `str`, mỗi giá trị có 2 dạng:

1. Là `0`
2. Là dạng dấu `$`, phía sau đó là giá tiền

Do đó, loại bỏ dấu `$`, sau đó chuyển sang `float`.

In [17]:
app_df['Price'] = app_df['Price'].replace(r'\$', '', regex=True)
app_df = app_df.astype({'Price': float})

## Chuyển dtype của cột "Last Updated" sang datetime

In [18]:
app_df['Last Updated'] = pd.to_datetime(app_df['Last Updated'], format="%B %d, %Y")

# Khám phá dữ liệu (tiếp tục)

## Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá trị được phân bố như thế nào?

Các cột có kiểu dữ liệu dạng số là `Reviews`, `Rating`, `Price`. Cột `Last Updated` đang có dtype là datetime nên xếp vào nhóm có kiểu dữ liệu dạng số luôn.

Với mỗi cột numeric sẽ tính tỉ lệ % giá trị thiếu (từ 0 đến 100), min, max rồi lưu kết quả vào dataframe `nume_col_profiles_df`. Dataframe này có 3 dòng là "missing_ratio", "min", "max"; và có 4 cột là `Reviews`, `Rating`, `Price`, `Last Updated`.

In [19]:
def missing_ratio(s):
    return s.isna().mean() * 100
nume_col_profiles_df = app_df[['Reviews', 'Rating', 'Price', 'Last Updated']] \
                        .agg([missing_ratio, pd.Series.min, pd.Series.max])
nume_col_profiles_df

Reviews     Rating  Price         Last Updated
missing_ratio         0.0  14.143657    0.0                  0.0
min                   0.0        NaN    0.0  2010-05-21 00:00:00
max            78158306.0        NaN  400.0  2018-08-08 00:00:00

Các giá trị này không có gì bất thường.

## Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?

Các cột có kiểu dữ liệu dạng phân loại dễ thấy nhất là cột `App`, `Category`, `Type`, `Content Rating`, `Genres`, `Current Ver`, `Android Ver`. 

Cột `Size` là dung lượng ứng dụng, nhưng có một giá trị là `Varies with device`, và các giá trị dung lượng cũng không cụ thể mà chỉ là 1 con số ước lượng nên có nhiều ứng dụng sẽ có cùng dung lượng. Vì thế xếp cột này vào dạng categorical.

Cột `Installs` thể hiện số lượt tải về ứng dụng, nhưng không phải là con số cụ thể mà chia làm các mốc về số lượt tải. Do đó xếp cột này vào dạng categorical.

Với mỗi cột categorical, ta tính tỉ lệ % giá trị thiếu (từ 0 đến 100), số lượng giá trị khác nhau (không xét giá trị thiếu), list các giá trị khác nhau (không xét giá trị thiếu). Kết quả lưu vào dataframe `cate_col_profiles_df`. Dataframe này có 3 dòng là `missing_ratio`, `num_diff_vals`, `diff_vals`; và có 9 cột là `App`, `Category`, `Type`, `Content Rating`, `Genres`, `Current Ver`, `Android Ver`, `Size`, `Installs`.

In [20]:
def missing_ratio(s):
    return s.isna().mean() * 100

def num_diff_vals(s):
    return s.dropna().nunique()

def diff_vals(s):
    return s.dropna().unique()

cate_col_profiles_df = app_df[ \
                    ['App', 'Category', 'Type', 'Content Rating', 'Genres', 'Current Ver', 'Android Ver', \
                    'Size', 'Installs']] \
                    .agg([missing_ratio, num_diff_vals, diff_vals])
cate_col_profiles_df

App  \
missing_ratio                                                0.0   
num_diff_vals                                               9660   
diff_vals      [Photo Editor & Candy Camera & Grid & ScrapBoo...   

                                                        Category  \
missing_ratio                                           0.009654   
num_diff_vals                                                 33   
diff_vals      [ART_AND_DESIGN, AUTO_AND_VEHICLES, BEAUTY, BO...   

                       Type  \
missing_ratio      0.009654   
num_diff_vals             2   
diff_vals      [Free, Paid]   

                                                  Content Rating  \
missing_ratio                                                0.0   
num_diff_vals                                                  6   
diff_vals      [Everyone, Teen, Everyone 10+, Mature 17+, Adu...   

                                                          Genres  \
missing_ratio                                           0.009654   
num_diff_vals                                                119   
diff_vals      [Art & Design, Art & Design;Pretend Play, Art ...   

                                                     Current Ver  \
missing_ratio                                           0.077235   
num_diff_vals                                               2831   
diff_vals      [1.0.0, 2.0.0, 1.2.4, Varies with device, 1.1,...   

                                                     Android Ver  \
missing_ratio                                           0.019309   
num_diff_vals                                                 33   
diff_vals      [4.0.3 and up, 4.2 and up, 4.4 and up, 2.3 and...   

                                                            Size  \
missing_ratio                                                0.0   
num_diff_vals                                                461   
diff_vals      [19M, 14M, 8.7M, 25M, 2.8M, 5.6M, 29M, 33M, 3....   

                                                        Installs  
missing_ratio                                                0.0  
num_diff_vals                                                 21  
diff_vals      [10,000+, 500,000+, 5,000,000+, 50,000,000+, 1...

Các giá trị này không có gì bất thường

# Đưa ra các câu hỏi có ý nghĩa cần trả lời

* Câu hỏi là gì?
* Nếu trả lời được câu hỏi thì sẽ có lợi ích gì?

# Tiền xử lý + phân tích dữ liệu để trả lời cho từng câu hỏi

# Nhìn lại quá trình làm đồ án

* Mỗi thành viên: Đã gặp những khó khăn gì? (Hay mọi
chuyện đều thuận lợi)
* Mỗi thành viên: Có học được gì hữu ích? (Hay không
học được gì)
* Nhóm: Nếu có thêm thời gian thì sẽ làm gì?

# Tài liệu tham khảo